File: scripts/dead_links.py
import os
import glob
import re

def check_for_bad_links(doc_path, links, image_links, data):
    reg = r'\[.+?\]\((\/.*?)\)'
    matches = re.findall(reg, data)
    for match in matches:        
        if match not in links:
            print(f"Dead link: {match} found in {doc_path}")

    # Check image links
    reg = r'!\[\]\((\/.*?)\)'
    matches = re.findall(reg, data)
    for match in matches:
        if match not in image_links:
            print(f"Dead image link: {match} found in {doc_path}")

def crawl_docs():
    file_paths = []
    links = []
    image_links = []

    base_directory = os.path.join(os.path.dirname(os.path.dirname(__file__)), "docs")
    for doc_path in glob.glob(base_directory + "/**/*.md", recursive=True):
        links.append(("/" + os.path.relpath(doc_path, base_directory)).replace(".md", "").replace("\\","/"))
        file_paths.append(doc_path)
    
    base_directory = os.path.join(os.path.dirname(os.path.dirname(__file__)), "docs", "public")
    for doc_path in glob.glob(base_directory + "/**/*", recursive=True):
        image_links.append(("/" + os.path.relpath(doc_path, base_directory)).replace("\\","/"))

    return file_paths, links, image_links

def main():
    file_paths, links, image_links = crawl_docs()

    for doc_path in file_paths:
        with open(doc_path, "r", encoding='utf-8') as f:
            check_for_bad_links(os.path.relpath(doc_path, os.getcwd()), links, image_links, f.read())

main()

File: scripts/dead_links.js
const fs = require('fs');
const glob = require('glob');
const path = require('path');

function checkForBadLinks(docPath, links, imageLinks, data) {
  let reg = /\[(.+?)\]\((\/.*?)\)/g;
  let matches;
  while ((matches = reg.exec(data)) !== null) {
    const match = matches[2];
    if (!links.includes(match)) {
      console.log(`Dead link: ${match} found in ${docPath}`);
    }
  }

  // Check image links
  reg = /!\[\]\((\/.*?)\)/g;
  while ((matches = reg.exec(data)) !== null) {
    const match = matches[1];
    if (!imageLinks.includes(match)) {
      console.log(`Dead image link: ${match} found in ${docPath}`);
    }
  }
}

function crawlDocs() {
  const filePaths = [];
  const links = [];
  const imageLinks = [];

  const baseDirectory = path.join(__dirname, '..', 'docs');
  for (const docPath of glob.sync(`${baseDirectory}/**/*.md`, { nodir: true })) {
    links.push(path.posix.join('/', path.relative(baseDirectory, docPath)).replace('.md', ''));
    filePaths.push(docPath);
  }

  const publicDirectory = path.join(baseDirectory, 'public');
  for (const docPath of glob.sync(`${publicDirectory}/**/*`, { nodir: true })) {
    imageLinks.push(path.posix.join('/', path.relative(publicDirectory, docPath)));
  }

  return { filePaths, links, imageLinks };
}

function main() {
  const { filePaths, links, imageLinks } = crawlDocs();

  for (const docPath of filePaths) {
    const data = fs.readFileSync(docPath, { encoding: 'utf-8' });
    checkForBadLinks(path.relative(process.cwd(), docPath), links, imageLinks, data);
  }
}

main();
